registry/consul/registry.go

```diff
@@ -0,0 +1,186 @@
+	// FIX: The change propagation (i.e. the channel communication) has to be somehow reworked.
+	// I currently "resolved" the issue by using buffered channel.
+	// Please note that if any service has the "delta" of instances (number of changes processed by `writeChanges()`
+	// method) greater than the channel buffer size it will deadlock. This is caused by the resolver which calls
+	// the `Instances()` method for each change which tries to read-lock the registry mutex.
+	// This is problematic in conjunction with the locking in the `writeChanges()` method which writes to this channel
+	// under write lock.
+	changesCh := make(chan registry.RegistryChange, 1024)
+	initializedCh := make(chan bool)
+	ctx, cancel := context.WithCancel(context.Background())
+	go r.watch(ctx, service, changesCh, initializedCh)
+	closeFn := func() {
+		close(changesCh)
+		cancel()
+	}
+	<-initializedCh
+	return changesCh, closeFn
+}
+
+func (r *consulRegistry) watch(ctx context.Context, service string, changesCh chan<- registry.RegistryChange, initializedCh chan<- bool) {
+	var consulService string
+	alias, ok := r.aliases[service]
+	if ok {
+		consulService = alias
+		logger.Infof("using alias (%s) to resolve service (%s)", alias, service)
+	}
+
+	if !ok {
+		consulService = service
+		logger.Warningf("no alias defined for service (%s)", service)
+	}
+
+	r.resolveService(service, consulService, changesCh)
+	logger.Infof("initialized registry for service (%s)", service)
+	initializedCh <- true
+
+	logger.Infof("watching changes for service (%s) every (%s)", service, r.refreshPeriod)
+	ticker := time.NewTicker(r.refreshPeriod)
+	for {
+		select {
+		case <-ticker.C:
+			r.resolveService(service, consulService, changesCh)
+
+		case <-ctx.Done():
+			logger.Infof("canceled watch for service (%s)", service)
+			return
+		}
+	}
+}
+
+func (r *consulRegistry) resolveService(service, consulService string, ch chan<- registry.RegistryChange) {
+	q := &consul.QueryOptions{
+		Datacenter: r.datacenter,
+	}
+
+	entries, _, err := r.client.Health().Service(consulService, "", true, q)
+	if err != nil {
+		logger.Errorf("failed listing consul health catalog for service (%s): %v", service, err)
+		return
+	}
+
+	instances := make([]registry.Instance, 0, len(entries))
+	for _, entry := range entries {
+		i := instance{
+			serviceName: service,
+			addr:        fmt.Sprintf("%s:%d", entry.Node.Node, entry.Service.Port),
+		}
+		instances = append(instances, i)
+	}
+
+	if len(instances) == 0 {
+		logger.Warningf("could not find any instances for service (%s)", service)
+	}
+
+	r.writeChanges(service, instances, ch)
+}
+
+func (r *consulRegistry) writeChanges(service string, newInstances []registry.Instance, ch chan<- registry.RegistryChange) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	oldInstances := r.services[service]
+	changed := false
+	for _, oldInstance := range oldInstances {
+		containsFunc := func(i registry.Instance) bool {
```

Comment:
Cannot this be defined on `registry.Instance` level?

Reply:
Well, it could be. But that might required making changes in the current `registry.Instance` and as of now, the goal is to merge these structure for all registries and delete the common interface that represents them. After these are merged, we can talk about introducing such new feature that we could be using for all the registries. I would leave it as is for now.

Reply:
Please create new issue for that, it would simplify the code and ad hoc function creation per each `watch`.

Reply:
Issue: https://github.com/moderntv/cadre/issues/46